src/ir/instance_fact.h

```diff
@@ -0,0 +1,119 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+
+#ifndef SRC_IR_INSTANCE_FACT_H_
+#define SRC_IR_INSTANCE_FACT_H_
+
+#include "src/ir/instantiator.h"
+#include "src/ir/noop_instantiator.h"
+
+// An InstanceFact is an instantation of some other fact that may be
+// initially not fully instantiated.
+//
+// Consider, for instance, the case of checks and claims on a ParticleSpec.
+// Checks and claims describe the check or claim they wish to perform on
+// AccessPaths rooted at the ParticleSpec. At some point, however, they must be
+// printed out as performing those claims and checks upon the access paths of
+// an actual instantiated particle.
+//
+// To do this, we create an InstanceFact upon the check or claim (or any
+// other instantiatable fact). The InstanceFact points at the original,
+// uninstantiated fact, and an Instantiator, which is just an object that
+// will give us a map from uninstantiated AccessPath roots to instantiated
+// AccessPath roots. Using the mapping provided by the instantiator, we can
+// transform details of the uninstantiated fact to be instantiated on the fly.
+// This prevents having to copy the data initially on the uninstantiated
+// fact, sidestepping the pointer ownership quandries and performance
+// penalties we may otherwise have to deal with.
+
+namespace raksha::ir {
+
+// This class represents an instantiation of a fact that was originally
+// stated about some uninstantiated root (such as a HandleConnectionSpec)
+// using an Instantiator.
+template<class T>
```

Comment:
I am not sure we gain too much value from this and other wrapper classes in this file. Isn't the instantiated class the same type as the uninstantiated class?